### ISSUE TITLE
`where` filter's operands order changes the result.

### ISSUE THREAD

**Issue by StefanBogdan** [semi-technologies/weaviate#1749]
If I run a `Get` query with some constraints I get different results depending on the order of the constraint order (but not always).  

I have these objects in weaviate:
```
{
  "data": {
    "Get": {
      "Videos": [
        {
          "duration": 57.8
        },
        {
          "duration": 109.3
        },
        {
          "duration": 177.8
        },
        {
          "duration": 51.8
        },
        {
          "duration": 45.5
        }
      ]
    }
  }
}
```

For example, if I run this query (for 15 <= `duration` <= 35 ):
**NOTE**: For this constraint there should be no results.
```
{
  Get {
    Videos(
      where: {
        operator: And
        operands: [
          {
            path: ["duration"],
            operator: GreaterThanEqual,
            valueNumber: 15
          },
          {
            path: ["duration"],
            operator: LessThanEqual,
            valueNumber: 35
          }
        ]
      }
    ) {
      duration
    }
  }
}
```
I get this results:
```
{
  "data": {
    "Get": {
      "Videos": [
        {
          "duration": 57.8
        },
        {
          "duration": 109.3
        }
      ]
    }
  }
}
```
But if I switch the order of the operands like this:
```
{
  Get {
    Videos(
      where: {
        operator: And
        operands: [
          {
            path: ["duration"],
            operator: LessThanEqual,
            valueNumber: 35
          },
          {
            path: ["duration"],
            operator: GreaterThanEqual,
            valueNumber: 15
          }
        ]
      }
    ) {
      duration
    }
  }
}
```
I get the correct result:
```
{
  "data": {
    "Get": {
      "Videos": []
    }
  }
}
```
Now if I have a constraint that has objects matching it, it works both ways:
```
{
  Get {
    Videos(
      where: {
        operator: And
        operands: [
          {
            path: ["duration"],
            operator: LessThanEqual,
            valueNumber: 60
          },
          {
            path: ["duration"],
            operator: GreaterThanEqual,
            valueNumber: 50
          }
        ]
      }
    ) {
      duration
    }
  }
}
```
Gets this result:
```
{
  "data": {
    "Get": {
      "Videos": [
        {
          "duration": 57.8
        },
        {
          "duration": 51.8
        }
      ]
    }
  }
}
```

and this query (switched operands):
```
{
  Get {
    Videos(
      where: {
        operator: And
        operands: [
          {
            path: ["duration"],
            operator: GreaterThanEqual,
            valueNumber: 50
          },
          {
            path: ["duration"],
            operator: LessThanEqual,
            valueNumber: 60
          }
        ]
      }
    ) {
      duration
    }
  }
}
```
has this result(same as the previous one):
```
{
  "data": {
    "Get": {
      "Videos": [
        {
          "duration": 57.8
        },
        {
          "duration": 51.8
        }
      ]
    }
  }
}
```

It seems that if one of the operands has zero results than it fails and the order has different results.

**Comment by etiennedi** [comment]
Thanks a lot for the detailed investigation and especially this statement:

> It seems that if one of the operands has zero results than it fails and the order has different results.

That will help a lot in narrowing this down. Thanks, I'll take it from here.

**Comment by etiennedi** [comment]
Hi @StefanBogdan, I tried to reproduce this using `v1.7.2`, `v1.8.0` and `v1.9.0`. Only `v1.7.2` ran into this issue for me. The others seemed fine, e.g:

![image](https://user-images.githubusercontent.com/8974479/145987531-fb3313d1-e02e-49e0-9af4-70fa99753004.png)

I think this bug may have already been fixed in `v1.8.0` where a lot of attention was paid to optimizing the filters and a lot of new tests were added. Could it be that some of the optimizations already fixed this? Can you confirm?

**Comment by StefanBogdan** [comment]
Hi @etiennedi , I did some test too, it seems that the bug is really fixed starting with  `v1.8.0`. The last version it persists is `v1.8.0-rc.3`(which was the latest at the time, should have included it in the Issue).

I am closing the issue.
